src/sqlparser/src/parser_v2/data_type.rs

```diff
@@ -0,0 +1,225 @@
+    identifier_non_reserved, keyword, literal_uint, object_name, precision_in_range, with_state,
+    TokenStream,
+};
+use crate::ast::{DataType, StructField};
+use crate::keywords::Keyword;
+use crate::tokenizer::Token;
+
+#[derive(Default, Debug)]
+struct DataTypeParsingState {
+    /// Since we can't distinguish between `>>` and `> >` in tokenizer, we need to handle this case in the parser.
+    /// When we want a [`>`][Token::Gt] but actually consumed a [`>>`][Token::ShiftRight], we set this to true.
+    /// When the value was true and we want a [`>`][Token::Gt], we just set this to false instead of really consume it.
+    remaining_close: Rc<RefCell<bool>>,
+}
+
+type StatefulStream<S> = Stateful<S, DataTypeParsingState>;
+
+/// Consume struct type definitions
+fn struct_data_type<S>(input: &mut StatefulStream<S>) -> PResult<Vec<StructField>>
+where
+    S: TokenStream,
+{
+    let remaining_close1 = input.state.remaining_close.clone();
+    let remaining_close2 = input.state.remaining_close.clone();
+
+    // Consume an abstract `>`, it may be the `remaining_close1` flag set by previous `>>`.
+    let consume_close = trace(
+        "consume_struct_close",
+        alt((
+            trace(
+                "consume_remaining_close",
+                move |input: &mut StatefulStream<S>| -> PResult<()> {
+                    if *remaining_close1.borrow() {
+                        *remaining_close1.borrow_mut() = false;
+                        Ok(())
+                    } else {
+                        fail(input)
+                    }
+                },
+            )
+            .void(),
+            trace(
+                "produce_remaining_close",
+                (
+                    Token::ShiftRight,
+                    move |_input: &mut StatefulStream<S>| -> PResult<()> {
+                        *remaining_close2.borrow_mut() = true;
+                        Ok(())
+                    },
+                )
+                    .void(),
+            ),
+            Token::Gt.void(),
+        )),
+    );
+
+    delimited(
+        Token::Lt,
+        separated(
+            1..,
+            trace(
+                "struct_field",
+                seq! {
+                    StructField {
+                        name: identifier_non_reserved,
+                        data_type: data_type_stateful,
+                    }
+                },
+            ),
+            Token::Comma,
+        ),
+        consume_close,
+    )
+    .context(StrContext::Label("struct_data_type"))
+    .parse_next(input)
+}
+
+/// Consume a data type definition.
+///
+/// The parser is the main entry point for data type parsing.
+pub fn data_type<S>(input: &mut S) -> PResult<DataType>
+where
+    S: TokenStream,
+{
+    #[derive(Debug, thiserror::Error)]
+    #[error("Unconsumed `>>`")]
+    struct UnconsumedShiftRight;
+
+    with_state::<S, DataTypeParsingState, _, _>(terminated(
+        data_type_stateful,
+        cut_err(trace(
```

Comment:
`cut_err` here seems unnecessary as the inner parser always returns `ErrMode::Cut`.